lib/acts_as_list/active_record/acts/list.rb

```diff
@@ -393,12 +393,27 @@ def shuffle_positions_on_intermediate_items(old_position, new_position, avoid_id
         def insert_at_position(position, raise_exception_if_save_fails=false)
           raise ArgumentError.new("position cannot be lower than top") if position < acts_as_list_top
           if invalid?
-            # `raise RecordInvalid` rather than causing a
-            # constraint violation on duplicate positions
+            # If you violate a uniqueness constraint on
+            # position_column, you get a postgres error,
+            # and the transaction rolls back.
+            # You can avoid this with a uniqueness validation (raising
+            # RecordNotValid) which you can catch and continue from.
             save! if raise_exception_if_save_fails
             return false
           end
           return set_list_position(position, raise_exception_if_save_fails) if new_record? && not_in_list?
+
+          # Cannot lock a record with unsaved changes, so save it first.
+          # If the changes are only to the position column, reset them, since we are overriding them.
+          self[position_column] = position_before_save if position_before_save_changed?
```

Comment:
Hi @DanielHeath, I think this method is starting to get a bit complicated. I've asked @swanandp to take a look at this code too when he's free just to make sure we're structuring things correctly. I don't want to get any of this wrong as it'll have far reaching consequences.

Reply:
Makes sense. 

I'm likely to deploy it to production for radiopaedia.org next week - will be able to report back whether it introduces any new runtime errors there (12 year old rails codebase, 26k lines of ruby in app, 700k daily visitors, uses acts_as_list in several places).

Reply:
This has been in production for a month or so with no issues (it's resolved our long-standing issues where we get non-sequential position values)